### PR TITLE
Feature 177 visibility refining

### LINKED ISSUE #177: Refining visibility
As a user I would like the visibility editor to be refined

### PR DESCRIPTION
**Pull request by Chi-Chu319**
resolves #177

### CODE REVIEW

File: src/utils/metaform-utils.ts [hunk 1]
```diff
@@ -394,6 +434,50 @@ namespace MetaformUtils {
...
+    fieldRule.or?.forEach(rule => fieldRuleScan(rule, match, foundFieldRules, fieldOptionMatch));
+  };
+
+  /**
+   * Checks field rules that matches
+   *
+   * @param fieldRule field rule
+   * @param match match
+   * @param fieldOptionMatch field option match
+   */
```
Comment (villejuutila) [inline src/utils/metaform-utils.ts:467]:
```suggestion
   * @param fieldOptionMatch field option match
     @return boolean whether field rule was matched
   */
```

File: src/utils/metaform-utils.ts [hunk 2]
```diff
@@ -11,6 +11,57 @@ import { FormContext } from "../types/index";
...
+  export const getMetaformSection = (form: Metaform, sectionId?: number): MetaformSection | undefined => {
+    if (sectionId === undefined) {
+      return undefined;
+    }
+
+    return form.sections?.[sectionId];
+  };
+
+  /**
+   * Gets metaform section
```
Comment (villejuutila) [inline src/utils/metaform-utils.ts:50]:
```suggestion
   * Gets metaform field
```

File: src/components/editor/metaform-editor-left-drawer.tsx [hunk 3]
```diff
@@ -49,7 +48,9 @@ const MetaformEditorLeftDrawer: React.FC<Props> = ({
   const renderFormTab = () => (
     <FormControl fullWidth>
       <Stack spacing={ 2 } padding={ 1 }>
-        <FormLabel>{ strings.draftEditorScreen.editor.form.versionInfo }</FormLabel>
+        <Typography variant="subtitle1">
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-left-drawer.tsx:51]: Not part of this PR but remove `padding={ 1 }` from line above this (50). Padding on this particular tab is higher than in others.

File: src/components/editor/metaform-editor-right-drawer-feature.tsx [hunk 4]
```diff
@@ -19,79 +22,95 @@ interface Props {
...
-  const [ columnType, setColumnType ] = useState<string>("");
-  const [ contextOption, setContextOption ] = useState<FormContext[]>([]);
+  const [ newColumnType, setNewColumnType ] = useState<MetaformTableColumnType>();
+  const [ selectedSection, setSelectedSection ] = useState<MetaformSection>();
+  const [ selectedField, setSelectedField ] = useState<MetaformField>();
+  const [ debounceTimerId, setDebounceTimerId ] = useState<NodeJS.Timeout>();
+
   /**
-   * Get title of current section
+   * Updates selected section, field
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:37]:
```suggestion
   * Updates selected section and field states
```

File: src/components/editor/metaform-editor-right-drawer-feature.tsx [hunk 5]
```diff
@@ -325,422 +320,435 @@ const MetaformEditorRightDrawerFeatureComponent: FC<Props> = ({
...
-            { strings.draftEditorScreen.editor.features.addSelectionField }
-          </Button>
-        </>
-      );
-    }
-  };
+   * Render multi-choice option edit
+   */
+  const renderMultiChoiceOptionEdit = (option: MetaformFieldOption, index: number) => (
+    <Stack spacing={ 2 } direction="row">
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:396]: Add key property.

File: src/components/editor/metaform-editor-right-drawer-feature.tsx [hunk 6]
```diff
@@ -325,422 +320,435 @@ const MetaformEditorRightDrawerFeatureComponent: FC<Props> = ({
...
+   *
+   * @param context context
+   * @param selectedContexts selected contexts
    */
-  const checkContextSettings = () => {
-    if (fieldIndex !== undefined && sectionIndex !== undefined) {
-      if (!pendingForm.sections![sectionIndex].fields![fieldIndex]) {
-        return;
+  const renderContextOption = (context: FormContext, selectedContexts: string[]) => (
+    <FormControlLabel
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:338]: Add key property

File: src/components/editor/metaform-editor-right-drawer-feature.tsx [hunk 7]
```diff
@@ -325,422 +320,435 @@ const MetaformEditorRightDrawerFeatureComponent: FC<Props> = ({
...
+  );
+
+  /**
+   * Renders table column edit
+   *
+   * @param column column
+   * @param index index
+   */
+  const renderTableColumnEdit = (column: MetaformTableColumn, index: number) => (
+    <Stack spacing={ 2 } direction="row">
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-feature.tsx:459]: Add key property

File: src/components/editor/metaform-editor-right-drawer-visibility.tsx [hunk 8]
```diff
@@ -23,306 +25,245 @@ const MetaFormRightDrawerVisibility: FC<Props> = ({
...
-  const [ requiredConditionEqualsValue, setRequiredConditionEqualsValue ] = React.useState<string>("");
-  const [ requiredConditionInfoField, setRequiredConditionInfoField ] = React.useState<boolean>(true);
-  const [ selectedComponent, setSelectedComponent ] = React.useState<string | undefined>(undefined);
-  const [ metaFormFieldSection, setMetaFormFieldSection ] = React.useState<MetaformField | MetaformSection>();
+  const [ selectedVisibleIf, setSelectedVisibleIf ] = React.useState<FieldRule | undefined>();
+  const [ visibleIfSource, setVisibleIfSource ] = React.useState<VisibilitySource>(VisibilitySource.NONE);
 
   /**
-   * Set values of Conditional field, switch and get selected component name
+   * Updates visibleIfSource section, field
```
Comment (villejuutila) [inline src/components/editor/metaform-editor-right-drawer-visibility.tsx:32]:
```suggestion
   * Updates visibleIf source section, field
```

**Review by villejuutila — CHANGES_REQUESTED** [review]
Added comments

**Review by villejuutila — APPROVED** [review]
Approved

File: src/localization/fi.json [hunk 9]
```diff
@@ -464,7 +467,14 @@
     "exportReplyXlsx": "Viety taulukkona"
   },
 
+  "formContext": {
+    "form": "Näkyy lomaketta täyttäessä",
+    "management": "Näkyy hallinnointi näkymässä",
+    "managementList": "Näkyy hallinnoinnin listaus näkymässä"
```
Comment (EVP1989) [inline src/localization/fi.json:473]:
```suggestion
    "form": "Näkyy lomaketta täyttäessä",
    "management": "Näkyy ammattilaiselle vastausnäkymässä",
    "managementList": "Näkyy ammattilaiselle vastauslistauksessa"
```

**Review by EVP1989 — APPROVED** [review]
Approved with one suggestion
